.github/copilot-instructions.md

```diff
@@ -0,0 +1,107 @@
+# Instructions for AIs
+
+This repository is .NET for Android.
+
+This is the main branch targeting .NET 10.
+
+## Nullable Reference Types
+
+When opting C# code into nullable reference types:
+
+* Add `#nullable enable` at the top of the file.
+
+* Don't *ever* use `!` to handle `null`!
+
+* Declare variables non-nullable, and check for `null` at entry points.
+
+* Use `throw new ArgumentNullException (nameof (parameter))` in `netstandard2.0` projects.
+
+* Use `ArgumentNullException.ThrowIfNull (parameter)` in Android projects that will be .NET 10+.
+
+* `[Required]` properties in MSBuild task classes should always be non-nullable with a default value.
+
+* Non-`[Required]` properties should be nullable and have null-checks in C# code using them.
+
+* For MSBuild task properties like:
+
+```csharp
+public string NonRequiredProperty { get; set; }
+public ITaskItem [] NonRequiredItemGroup { get; set; }
+
+[Output]
+public string OutputProperty { get; set; }
+[Output]
+public ITaskItem [] OutputItemGroup { get; set; }
+
+[Required]
+public string RequiredProperty { get; set; }
+[Required]
+public ITaskItem [] RequiredItemGroup { get; set; }
+```
+
+Fix them such as:
+
+```csharp
+public string? NonRequiredProperty { get; set; }
+public ITaskItem []? NonRequiredItemGroup { get; set; }
+
+[Output]
+public string? OutputProperty { get; set; }
+[Output]
+public ITaskItem []? OutputItemGroup { get; set; }
+
+[Required]
+public string RequiredProperty { get; set; } = "";
+[Required]
+public ITaskItem [] RequiredItemGroup { get; set; } = [];
+```
+
+If you see a `string.IsNullOrEmpty()` check:
+
+```csharp
+if (!string.IsNullOrEmpty (NonRequiredProperty)) {
+    // Code here
+}
+```
+
+Convert this to:
+
+```csharp
+if (NonRequiredProperty is { Length: > 0 }) {
```

Comment:
never seen this kind of code before? Is this new?

Reply:
The problem is `string.IsNullOrEmpty()` isn't doesn't have NRT annotations in `netstandard2.0`, but the new "pattern matching" works instead.

There are a lot of ways to use it shown here:

* https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#property-pattern

I have not tried all these yet...

Reply:
Another option is: 
https://github.com/dotnet/android/blob/21c413195e300b6440eb437dade4f3a114e795f7/src/Xamarin.Android.Build.Tasks/Utilities/MavenExtensions.cs#L24-L27

But either is fine.

Reply:
Yeah `netstandard2.0` is missing the `[NotNullWhen (true)]`.

Reply:
Can we move to netstandard2.1 ?

Reply:
It doesn't look like 2.1 is supported by .NET framework:
* https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-1